### DISCUSSION TITLE
[REQUEST] Native ZBOSS Zigbee stack support for Zigbee2Tasmota on ESP32-C6?

### DISCUSSION THREAD

**Discussion by Hedda** [arendst/Tasmota#20045] [Ideas]
@s-hadinger Do you have any plans on looking into native ZBOSS Zigbee stack support for Zigbee2Tasmota on [ESP32-C6](https://github.com/arendst/Tasmota/discussions/19350) built-in radio? 

Not sure if any use to help achieve native ZBOSS Zigbee support in Tasmota, but FYI, as an alternative to the upstream [ESP-IDF Zigbee library](https://components.espressif.com/components/espressif/esp-zigbee-lib), it looks like @P-R-O-C-H-Y is planning on porting the [ESP-IDF Zigbee library](https://components.espressif.com/components/espressif/esp-zigbee-lib) and ZBOSS libraries to be available as Arduino library/libraries for ESP32. Check it out here:

https://github.com/espressif/arduino-esp32/issues/8807

https://github.com/espressif/esp32-arduino-lib-builder/issues/142

https://github.com/espressif/arduino-esp32/pull/8844

https://github.com/espressif/esp32-arduino-lib-builder/pull/146

Would it be possible to make [Zigbee2Tasmota (Z2T)](https://tasmota.github.io/docs/Zigbee/) compatible with a serial interface for Espressif's "esp-zboss-lib" (ESP ZBOSS 3.0 Libraries) or other ZBOSS Zigbee library to allow using [Zigbee2Tasmota (Z2T)](https://tasmota.github.io/docs/Zigbee/) as a gateway with no other hardware and/or serial-server on ESP32-C6 for Serial-over-IP pass-through proxy (similar to ser2net) to enable TCP streaming of the serial communication bridge to allow it to be used as a network-attached remote Zigbee Coordinator so that the Home Assistant's [ZHA integration](https://www.home-assistant.io/integrations/zha) can connect to it via a socket (socat) using the experimental [zigpy-zboss](https://github.com/kardia-as/zigpy-zboss)?

An open question asked to Espressif that might be a showstopper for  [Zigbee2Tasmota (Z2T)](https://tasmota.github.io/docs/Zigbee/) compatible is if Espressif's current ZBOSS Zigbee libraries for ESP32-C6 currently only support RCP (Radio Co-Processor) mode and not Zigbee NCP (Network Co-Processor) host? Otherwise you would have to also implement RCP (Radio Co-Processor) mode support, however, that might not be impossible if supported upstream, or?

https://github.com/espressif/esp-zboss-lib/issues/4

https://github.com/espressif/esp-zigbee-sdk/issues/124

For reference, Tasmota's Zigbee2Tasmota (Z2T) already supports Texas Instruments ZNP and Silicon Labs EZSP Zigbee stacks/interfaces with respective Zigbee Coordinator adapter running Zigbee firmware in NCP (Network Co-Processor) mode:

https://tasmota.github.io/docs/Zigbee/

[Espressif of ESP32 fame](https://en.wikipedia.org/wiki/ESP32) joined [ZOI (ZBOSS Open Initiative)](https://dsr-zoi.com/) more than a year ago and has released ESP ZBOSS 3.0 binary libraries for ESP-IDF supporting the Zboss Zigbee 3.0 stack for [ESP32-C6](https://www.espressif.com/en/products/socs/esp32-c6) and [ESP32-H2](https://www.espressif.com/en/products/socs/esp32-h2) SoCs which both features embedded 802.15.4 radio for Zigbee (and Thread) support, (with the major difference being that the ESP32-H2  hardware does not have WiFi and only one core core)

https://github.com/espressif/esp-zigbee-sdk

https://github.com/espressif/esp-zboss-lib

https://components.espressif.com/components/espressif/esp-zboss-lib

FYI, Espressif has also released a couple of development kits based on ESP32-H2 as reference hardware for such and similar solutions:

https://www.cnx-software.com/2023/06/20/espressif-esp-thread-border-router-board-combines-esp32-h2-esp32-c3-wireless-chips/

* https://github.com/espressif/esp-thread-br
* https://docs.espressif.com/projects/esp-thread-br/en/latest/esp32/
* https://www.aliexpress.us/item/1005005688193617.html

![image](https://github.com/kardia-as/zigpy-zboss/assets/6320001/fbee9955-a376-43c4-b442-4f71fa7177e7)

That is a new ESP32-H2 + ESP32-C3 combo development board kit marketed as ESP Thread Boarder Router / Zigbee Gateway with an optional Ethernet daughter-board/sub-board sold separately (which would normally be recommended as [streaming serial communication over WiFi does not work well](https://www.home-assistant.io/integrations/zha#warning-about-wi-fi-based-zigbee-to-serial-bridgesgateways)), although that Ethernet daughter-board/sub-board is based on WIZnet W5500 which uses a non-standard SPI interface instead of RMII so might [not yet be supported in Tasmota](https://github.com/arendst/Tasmota/discussions/18207))(?).

* https://github.com/espressif/esp-idf/tree/master/examples/openthread/ot_br

* https://openthread.io/guides/border-router/espressif-esp32

An earlier two-board ESP Thread Border Router / Zigbee Gateway solution that combined separate ESP32-C6 and ESP32-H2 boards was previously available as an example project for reference to achieve the same result:

* https://github.com/espressif/esp-idf/tree/master/examples/openthread/ot_br
* https://openthread.io/guides/border-router/espressif-esp32

![image](https://github.com/kardia-as/zigpy-zboss/assets/6320001/4793c891-19c0-4ae1-9a97-56076b9a6357)

With either of those hardware combinations it should be possible to make a network-attached "Zigbee Coordinator" adapter products based on [Tasmota](https://tasmota.github.io/docs/Zigbee/) (or [ESPHome](https://esphome.io/) ) similar to [TubesZB Zigbee Gateways](https://github.com/tube0013/tube_gateways) with ESP32-C6 or ESP32-H2 that simply tunnel/passthrough the serial connection to the serial interface for Espressif's "esp-zboss-lib" (ESP ZBOSS 3.0 Libraries) from Home Assistant's [ZHA integration](https://www.home-assistant.io/integrations/zha).

# ESP ZBOSS 3.0 Libraries

This repository contains binary libraries supporting the Zboss Zigbee 3.0 stack for ESP32 series chips.

Packages from this repository are uploaded to [Espressif’s component service](https://components.espressif.com/). You can add them to your project via idf.py add-dependency

More information about idf-component-manager can be found in [Espressif API guide](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-guides/tools/idf-component-manager.html) or [PyPi registry](https://pypi.org/project/idf-component-manager/).

There now looks to be some Zigbee examples in Espressif’s IoT Development Framework (ESP-IDF):

* https://components.espressif.com/component/espressif/esp-zboss-lib
* https://www.espressif.com/en/products/sdks/esp-idf
* https://github.com/espressif/esp-idf
* https://github.com/espressif/esp-idf/tree/master/examples/zigbee
  * https://github.com/espressif/esp-idf/tree/master/examples/zigbee/esp_zigbee_gateway
  * https://github.com/espressif/esp-idf/tree/master/examples/zigbee/esp_zigbee_rcp
  * https://github.com/espressif/esp-idf/tree/master/examples/zigbee/light_sample

Those examples include; Zigbee coordinator ("light coordinator"), "Zigbee RCP (radio co-processor"), and "Zigbee gateway":

* https://github.com/espressif/esp-idf/tree/master/examples/zigbee/light_sample/light_coordinator

"light_coordinator is a light coordinator example demonstrating Zigbee Coordinator role. It provides a formation of the Zigbee network. It runs on an 802.15.4 SoC like ESP32-H2. For more details to see the example readme file."

* https://github.com/espressif/esp-idf/tree/master/examples/zigbee/esp_zigbee_gateway

* https://github.com/espressif/esp-idf/tree/master/examples/zigbee/esp_zigbee_rcp

Wondering if "RPC" works similarly to Silicon Labs RPC which works like a dumb radio requiring external ESP ZBOSS 3.0 libraries?

* https://components.espressif.com/component/espressif/esp-zboss-lib

"This test code shows how to configure Zigbee rcp (radio co-processor) device. Rcp doesn't function alone, it needs to work together with Zigbee gateway (see esp_zigbee_gateway example)"

"After rcp starts up, it will send its own MAC ieee address and Zigbee stack version number to the Zigbee gateway and start working together with Zigbee gateway via UART communication"

ESP32-H2 was apparently early on even certified as a “Zigbee Compliant Platform” by the CSA:

* https://www.espressif.com/en/news/ESP32-H2_Thread_and_Zigbee_certified

* https://www.espressif.com/sites/default/files/ESP32-H2%20Zigbee%20Compliant%20Platform%20Certification.pdf

Type of Device: Zigbee Compliant Platform
Zigbee PRO Feature Set (2017)
Manufacturer: Espressif Systems (Shanghai) Co., Ltd.
Model Identification: ESP32-H2
Firmware Version: V1.0
Hardware Version: V1.0
Certification Date: October 20, 2021
Certification ID Number: ZIG21030ZCP27315-24

# External links, documentation, and other development references

- [ZBOSS NCP Serial Protocol (v1.5) prepared by DSR Corporation for ZOI](https://cloud.dsr-corporation.com/index.php/s/BAn4LtRWbJjFiAm)
- https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/index.html - Specifically see the ZBOSS NCP sample, Zigbee CLI examples, and the ZBOSS NCP Host user guide.
  - https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/samples/zigbee/ncp/README.html
  - https://infocenter.nordicsemi.com/index.jsp?topic=%2Fsdk_tz_v4.1.0%2Fzigbee_only_examples.html
  - https://developer.nordicsemi.com/nRF_Connect_SDK/doc/zboss/3.6.0.0/zboss_ncp_host_intro.html
    - https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/protocols/zigbee/architectures.html#ug-zigbee-platform-design-ncp-details
    - https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/samples/zigbee/ncp/README.html#zigbee-ncp-sample
    - https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/protocols/zigbee/tools.html#ug-zigbee-tools-ncp-host
    - https://developer.nordicsemi.com/nRF_Connect_SDK/doc/zboss/3.6.0.0/zboss_ncp_host.html
    - https://developer.nordicsemi.com/nRF_Connect_SDK/doc/zboss/3.11.2.1/zboss_ncp_host_intro.html
    - https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/protocols/zigbee/tools.html
      - https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/protocols/zigbee/tools.html#ug-zigbee-tools-ncp-host
        - [ZBOSS NCP Host (v2.2.1) source code](https://developer.nordicsemi.com/Zigbee/ncp_sdk_for_host/ncp_host_v2.2.1.zip)
- https://github.com/kardia-as/zigpy-zboss
  - https://github.com/zigpy/zigpy/issues/394 - Previous development discussion about ZBOSS radio library for zigpy.
    - https://github.com/zigpy/zigpy/discussions/595 - Reference collections for Zigbee Stack and related dev docks
      - https://github.com/MeisterBob/zigpy_nrf52 - Other attemt at making a zigpy library for nFR52
         - https://gist.github.com/tomchy/04ac4ff78d6e117d33ab92d9cc1de779 - Another attemt at making a zigpy controller for nFR

**Comment by Jason2866** [comment]
Thx for collecting and providing Infos. We had already a quick look at the espressif zigbee stuff. To early to say anything more. It is on our list to check ;-)

**Reply by s-hadinger** [reply]
I do have an Espressif C6/H2 board to allow some testing. However I'm hesitant to invest time in Zigbee support, considering there is already plenty of options with ZNP/EZSP. I have the feeling that with C6/H2, we should now target Thread support

**Reply by Hedda** [reply]
IMHO, while Matter over Thread might be the future it will take many years to catch up with where Zigbee devices are today, both feature-wise and market adoption. There are already thousands of inexpensive Zigbee devices on the market today, and also note that the Matter device type is still limited and does not allow custom manufacturer clusters and attributes, so it will take years for the new Matter standard to be extended and mature in a way to compete with the flexibility of the existing Zigbee specifications.

**Reply by Jason2866** [reply]
Tasmota supports zigbee, and will do in future. The two Tasmota supported zigbee chip familys are widley available and have good support. Tasmotas goal is not to support every chip especially when it brings no new features. There is a higher chance that the new feature Thread will be added, than adding support for another zigbee chip.
The espressif way has an major disadvantage against current zigbee solutions. Everything is closed source and just precompiled libs are provided.
So at the moment we decided NOT to support espressif zigbee in Tasmota.

**Reply by Hedda** [reply]
> Tasmotas goal is not to support every chip especially when it brings no new features. 

Yeah I totally understand that, however it is the only Zigbee radio that is built-into ESP32 chips, and I guess that can be called a feature.

> There is a higher chance that the new feature Thread will be added, than adding support for another zigbee chip.

Indeed, and that is also very understandable. Especially with the fact that if using OpenThread then it is the same CLI/API on all hardware.

**Comment by Hedda** [comment]
> The espressif way has an major disadvantage against current zigbee solutions. Everything is closed source and just precompiled libs are provided.

I believe that that is only partially true. Yes sadly Espressif has chosen to make their initial implementation not so open-source friendly, however on the positive note it has potential because Espressif at least chose to go with the ZBOSS Zigbee stack which I understand is very open-source upstream, and since the same the ZBOSS Zigbee stack has also been adopted other by several others, including Nordic Semiconductor and Telink.

https://dsr-zoi.com

https://www.dsr-zboss.com

Tip is to check out the "nRF52840 Dongle" (nRF52840 SoC based USB dongle for developers and tinkerers) made by Nordic Semiconductor.

So an option could be to develop using Nordic Semi's nRF5 series hardware (nRF52840 SoC based development boards) and documentation instead, then port that, as I understand that Nordic Semiconductor has both great developer documentation as well as inexpensive hardware designed for developers that are readily available. 

https://www.nordicsemi.com/Products/Development-hardware/nrf52840-dongle

https://www.nordicsemi.com/Products/nRF52840

https://www.nordicsemi.com/Products/Development-hardware/nrf52840-dk

https://docs.nordicsemi.com/category/zigbee

https://www.nordicsemi.com/Products/Development-software/nrf5-sdk-for-thread-and-zigbee

https://infocenter.nordicsemi.com/index.jsp

PS: There are also a few ESP32-C3 boards out there that feature Telink TLSR8258 multi-protocol wireless SoC with ZBOSS Zigbee Stack, (like example the [LilyGO T-Zigbee board](https://www.cnx-software.com/2022/04/27/10-t-zigbee-board-combines-esp32-c3-and-tlsr8258-for-zigbee-3-0-wifi-and-ble-connectivity/)), however, I believe that Telink do not provide as good documentation as Nordic Semiconductor. There are also radio modules with Telink TLSR8258  available from Ebyte ([E180-Z5812SP](https://www.ebyte.com/en/product-view-news.html?id=1376))

**Comment by Jason2866** [comment]
> Yeah I totally understand that, however it is the only Zigbee radio that is built-into ESP32 chips, and I guess that can be called a feature.

Probably this is a problem, like it is currently when using BLE and WiFi together. It is not working perfect since the HF hardware sharing is problematic. So for a good solution it is better to have a dedicated zigbee chip, just doing the zigbee job.

Time will show if well working solutions with the C6 are possible where wifi and zigbee is handled from the C6.
Project Tasmota wont be the one who is gonna try first or ever.

Thread has the same problem, so it might be necessary to use the C6 additionally to an existing ESP.

About existing hardware boards, just soldering chips on a PCB makes no working solution.

**Comment by Jason2866** [comment]
> I believe that that is only partially true. Yes sadly Espressif has chosen to make their initial implementation not so open-source friendly, however on the positive note it has potential because Espressif at least chose to go with the ZBOSS Zigbee stack which I understand is very open-source upstream, and since the same the ZBOSS Zigbee stack has also been adopted other by several others, including Nordic Semiconductor and Telink.

Why should the Open Source community support such a unfriendly decision? Honestly I don't care which way the choose, when the have the chance to provide the source code.

**Reply by Hedda** [reply]
Texas Instruments and Silicon Labs are also proprietary Zigbee stacks/solutions that are mostly closed source, they all more or less provide CLIs and APIs interfaces so that third parties can utilise them. ZBOSS is the one that is the most open and the only one that is hardware-independent.
